### PR TITLE
Added Sessions Functionality

### PR DESCRIPTION
**Pull request by L0Lmaker**
*Issue #, if available:*
https://bitquill.atlassian.net/jira/software/c/projects/AN/boards/36?modal=detail&selectedIssue=AN-980

*Description of changes:*
Added Sessions Functionality

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gremlin-go/driver/client.go [hunk 1]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
 		transporterType: settings.TransporterType,
 		connection:      conn,
 	}
+	// TODO: PoolSize must be 1 on session mode. Implement after AN-980
```
Comment (lyndonbauto) [inline gremlin-go/driver/client.go]: Where is this from? Not saying it doesn't make sense but I don't see this lock in gremlin-python

File: gremlin-go/driver/client.go [hunk 2]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+	var processor string
+	var op string
+	switch message.(type) {
+	case bytecode:
+		client.logHandler.logf(Debug, bytecodeReceived, message)
+		op = "bytecode"
+		processor = "traversal"
+	case string:
+		client.logHandler.logf(Debug, stringReceived, message)
+		// TODO: implement after bindings (AN-1018).
```
Comment (lyndonbauto) [inline gremlin-go/driver/client.go]: Remove AN-#

File: gremlin-go/driver/client.go [hunk 3]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+		op = "eval"
+	default:
+		return nil, errors.New("message must either be a string or bytecode, neither was passed")
+	}
+	// If session
+	if client.session != "" {
+		args["session"] = client.session
+		processor = "session"
+	}
+	// TODO: Get connection from pool after AN-982
```
Comment (lyndonbauto) [inline gremlin-go/driver/client.go]: Remove AN-#

File: gremlin-go/driver/client.go (hunk 1 above)
Comment (L0Lmaker) [reply gremlin-go/driver/client.go]: Gremlin-Python has this block in the client constructor
```python
        if self._session_enabled:
            if pool_size is None:
                pool_size = 1
            elif pool_size != 1:
                raise Exception("PoolSize must be 1 on session mode!")
```

File: gremlin-go/driver/driverRemoteConnection.go [hunk 4]
```diff
@@ -80,14 +85,38 @@ func NewDriverRemoteConnection(
...
 
 	return &DriverRemoteConnection{client: client}, nil
 }
 
 // Close closes the DriverRemoteConnection.
 func (driver *DriverRemoteConnection) Close() error {
-	return driver.client.Close()
+	// If DriverRemoteConnection has spawnedSessions then they must be closed as well.
+	if len(driver.spawnedSessions) > 0 {
+		driver.client.logHandler.logger.Logf(Info, "closing spawned sessions from DriverRemoteConnection with url '%s'", driver.client.url)
```
Comment (lyndonbauto) [inline gremlin-go/driver/driverRemoteConnection.go]: I think you should be using our logger in the other fashion with the resources.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 5]
```diff
@@ -80,14 +85,38 @@ func NewDriverRemoteConnection(
...
+			}
+		}
+		driver.spawnedSessions = []*DriverRemoteConnection{}
+	}
+
+	if driver.IsSessionBound() {
+		driver.client.logHandler.logger.Logf(Info, "closing DriverRemoteConnection with url '%s' with session '%s'", driver.client.url, driver.client.session)
+	} else {
+		driver.client.logHandler.logger.Logf(Info, "closing DriverRemoteConnection with url '%s'", driver.client.url)
+	}
```
Comment (lyndonbauto) [inline gremlin-go/driver/driverRemoteConnection.go]: I think you should be using our logger in the other fashion with the resources.

File: gremlin-go/driver/client.go [hunk 6]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+	var op string
+	switch message.(type) {
+	case bytecode:
+		client.logHandler.logf(Debug, bytecodeReceived, message)
+		op = "bytecode"
+		processor = "traversal"
+	case string:
+		client.logHandler.logf(Debug, stringReceived, message)
+		// TODO: implement after bindings (AN-1018).
+		// args['bindings'] = bindings (Add Argument to func)
```
Comment (xiazcy) [inline gremlin-go/driver/client.go]: Just to note, this is being added in my bindings branch.

File: gremlin-go/driver/client.go [hunk 7]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+	args := map[string]interface{}{
+		"gremlin": message,
+		"aliases": map[string]interface{}{
+			"g": client.traversalSource,
+		},
+	}
+	var processor string
+	var op string
+	switch message.(type) {
+	case bytecode:
```
Comment (xiazcy) [inline gremlin-go/driver/client.go]: If we are adding a `bytecode` type check here, is `submitBytecode` still needed?

File: gremlin-go/driver/request.go [hunk 8]
```diff
@@ -75,3 +75,12 @@ func makeBasicAuthRequest(auth string) (req request) {
 		},
 	}
 }
+
+func makeRequest(op string, processor string, args map[string]interface{}) (req request) {
```
Comment (xiazcy) [inline gremlin-go/driver/request.go]: Again, with this added function, would the above ones be considered redundant?

File: gremlin-go/driver/driverRemoteConnection.go (hunk 5 above)
Comment (L0Lmaker) [reply gremlin-go/driver/driverRemoteConnection.go]: Im not sure I understand what you mean

Comment (xiazcy) [reply gremlin-go/driver/driverRemoteConnection.go]: Check out `logger.go` and the `resources` folder and see how log messages are structured (note that even though the type is called `errorKey`, we are logging state, it's just a bad name for now)

File: gremlin-go/driver/client.go [hunk 9]
```diff
@@ -33,6 +34,8 @@ type ClientSettings struct {
 	Language        language.Tag
 	AuthInfo        *AuthInfo
 	TlsConfig       *tls.Config
+	session         string
+	closed          bool
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: I suspect some oversight here.

Firstly, double check if a user of gremlingo is expected to pass a session into a `Client`. If yes, then it needs to be exported.

`closed` should not exist here in either context, however.

It would also be a good idea to take a bit of time to understand how the constructor of the client interacts with `ClientSettings`.

File: gremlin-go/driver/client.go [hunk 10]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+		_, err := client.closeSession()
+		if err != nil {
+			return err
+		}
+	}
+	client.logHandler.logger.Logf(Info, "Closing Client with url %s", client.url)
 	return client.connection.close()
 }
 
+func (client *Client) IsClosed() bool {
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: I don't see value here in increasing API surface area for this.

Is there a reason this method needs to exist instead of just invoking `client.closed`?

File: gremlin-go/driver/client.go [hunk 11]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+		return nil
+	}
+	// If it is a Session, call closeSession
+	if client.session != "" {
+		_, err := client.closeSession()
+		if err != nil {
+			return err
+		}
+	}
+	client.logHandler.logger.Logf(Info, "Closing Client with url %s", client.url)
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: There shouldn't be entrance into the logHandler's internals since the logHandler's API itself provides what you need.

Following that API would also fix a minor mistake here, which is a hardcoded error string instead of one contained in the resources bundle accessed via a `key` for localization.

File: gremlin-go/driver/client.go [hunk 12]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+func (client *Client) IsClosed() bool {
+	return client.closed
+}
+
 // Submit submits a Gremlin script to the server and returns a ResultSet.
-func (client *Client) Submit(traversalString string) (ResultSet, error) {
-	// TODO AN-982: Obtain connection from pool of connections held by the client.
-	client.logHandler.logf(Debug, submitStartedString, traversalString)
-	request := makeStringRequest(traversalString, client.traversalSource)
+func (client *Client) Submit(message interface{}) (ResultSet, error) {
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: This change in API concerns me for a few reasons and I want to make sure it's correct.

1. This makes it error prone as the input is no longer type enforced
2. This is a public-facing function, and removing the type parameter makes it confusing to use
3. There is a distinction between `bytecode` and `string` types for logic within this method, but there is another method in this struct, `submitBytecode` meant to handle bytecode. Why was it not used instead? Why is the method meant to handle strings now handling both?

File: gremlin-go/driver/client.go [hunk 13]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+	default:
+		return nil, errors.New("message must either be a string or bytecode, neither was passed")
+	}
+	// If session
+	if client.session != "" {
+		args["session"] = client.session
+		processor = "session"
+	}
+	// TODO: Get connection from pool
+	client.logHandler.logf(Debug, "processor='%s', op='%s', args='%s'", processor, op, args)
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: I don't think this is going to work since you're passing a hardcoded string instead of a key, which it will use to search for a message and fail.

File: gremlin-go/driver/client.go [hunk 14]
```diff
@@ -93,3 +145,10 @@ func (client *Client) submitBytecode(bytecode *bytecode) (ResultSet, error) {
 	request := makeBytecodeRequest(bytecode, client.traversalSource)
 	return client.connection.write(&request)
 }
+
+func (client *Client) closeSession() (ResultSet, error) {
+	message := makeRequest("close", "session", map[string]interface{}{
+		"session": client.session,
+	})
+	return client.connection.write(&message)
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: Looking at this makes me think that tracking `closed` isn't necessary in the `client`.

The connection, or whatever lower layer where appropriate should no-op so that the client does not care.

File: gremlin-go/driver/client.go [hunk 15]
```diff
@@ -71,19 +78,64 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+		client.logHandler.logf(Debug, bytecodeReceived, message)
+		op = "bytecode"
+		processor = "traversal"
+	case string:
+		client.logHandler.logf(Debug, stringReceived, message)
+		// TODO: implement after bindings
+		// args['bindings'] = bindings (Add Argument to func)
+		processor = ""
+		op = "eval"
+	default:
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: This entire part regarding the args, processor, op, etc. don't match 1 for 1 with our current existing implementation even outside of when it is needed to handle sessions.

Is this intentional, and more importantly, correct?

File: gremlin-go/driver/driverRemoteConnection.go [hunk 16]
```diff
@@ -80,14 +85,38 @@ func NewDriverRemoteConnection(
...
 // Close closes the DriverRemoteConnection.
 func (driver *DriverRemoteConnection) Close() error {
-	return driver.client.Close()
+	// If DriverRemoteConnection has spawnedSessions then they must be closed as well.
+	if len(driver.spawnedSessions) > 0 {
+		driver.client.logHandler.logger.Logf(Info, "closing spawned sessions from DriverRemoteConnection with url '%s'", driver.client.url)
+		for _, session := range driver.spawnedSessions {
+			err := session.Close()
+			if err != nil {
+				return err
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: This close is going to fail and not close the rest of the resources.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 17]
```diff
@@ -80,14 +85,38 @@ func NewDriverRemoteConnection(
...
+			}
+		}
+		driver.spawnedSessions = []*DriverRemoteConnection{}
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: Why is this necessary? 

If an instance of something is closed, it is not meant to be used again.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 18]
```diff
@@ -80,14 +85,38 @@ func NewDriverRemoteConnection(
...
+	if driver.IsSessionBound() {
+		driver.client.logHandler.logger.Logf(Info, "closing DriverRemoteConnection with url '%s' with session '%s'", driver.client.url, driver.client.session)
+	} else {
+		driver.client.logHandler.logger.Logf(Info, "closing DriverRemoteConnection with url '%s'", driver.client.url)
+	}
+	err := driver.client.Close()
+	if err != nil {
+		return err
+	}
+	driver.client.closed = true
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: I don't think you should modify the state tracker of a different type outside of the type.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 19]
```diff
@@ -100,6 +129,41 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
 	return driver.client.submitBytecode(bytecode)
 }
 
+// IsSessionBound returns True if a DriverRemoteConnection is a Session
+func (driver *DriverRemoteConnection) IsSessionBound() bool {
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: Is this necessary to be exposed?

File: gremlin-go/driver/driverRemoteConnection.go [hunk 20]
```diff
@@ -100,6 +129,41 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
...
+	return driver.client.session != ""
+}
+
+// CreateSession generates a new Session. sessionId stores the optional UUID param. It can be used to create a session with a specific UUID.
+func (driver *DriverRemoteConnection) CreateSession(sessionId ...string) (*DriverRemoteConnection, error) {
+	driver.client.logHandler.logger.Log(Info, "creating session based connection")
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: Ditto here regarding the logger.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 21]
```diff
@@ -100,6 +129,41 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
...
+			settings.Session = uuid.New().String()
+		}
+	}
+	connection, err := NewDriverRemoteConnection(driver.client.url, sessionHandler)
+	if err != nil {
+		return nil, err
+	}
+	driver.spawnedSessions = append(driver.spawnedSessions, connection)
+	return connection, nil
+}
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go:148]: This entire portion is a bit confusing to me.

It doesn't seem like the session is actually being used by any clients, whether it be being created with one, or having CreateSession be invoked.

The distinction between adding a session to an existing DRC and creating a new one with a session are different use cases but it's not very obvious what is for what, as you can do both. Creating a session from an existing DRC is also confusing because it creates an instance of the same type.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/client.go (hunk 9 above)
Comment (lyndonbauto) [reply gremlin-go/driver/client.go]: Session can optionally be passed to a client, so I think it should be exported.

File: gremlin-go/driver/client.go (hunk 10 above)
Comment (lyndonbauto) [reply gremlin-go/driver/client.go]: I think just client.closed is okay - I can't see a use case outside of internal mechanics.

File: gremlin-go/driver/client.go (hunk 12 above)
Comment (lyndonbauto) [reply gremlin-go/driver/client.go]: Python does it this way but I think Simon has a point here.

I think we should probably retain separate functions here.

File: gremlin-go/driver/client.go (hunk 7 above)
Comment (lyndonbauto) [reply gremlin-go/driver/client.go]: Let's keep submitBytecode, we shouldn't allow a user to submit bytecode directly.

Comment (lyndonbauto) [reply gremlin-go/driver/client.go]: It's an internal type.

File: gremlin-go/driver/client.go (hunk 10 above)
Comment (L0Lmaker) [reply gremlin-go/driver/client.go:101]: The reason I added this was so closed can be read but not edited. Exposing closed would mean that the state could changed by the user. Is there an alternate way to do this?

File: gremlin-go/driver/driverRemoteConnection.go (hunk 16 above)
Comment (L0Lmaker) [reply gremlin-go/driver/driverRemoteConnection.go]: How should we handle this case ideally?

File: gremlin-go/driver/driverRemoteConnection.go (hunk 17 above)
Comment (L0Lmaker) [reply gremlin-go/driver/driverRemoteConnection.go]: Is it safe to assume the connection is not re-opened?

File: gremlin-go/driver/driverRemoteConnection.go (hunk 19 above)
Comment (L0Lmaker) [reply gremlin-go/driver/driverRemoteConnection.go]: not really, will change

File: gremlin-go/driver/client.go (hunk 14 above)
Comment (L0Lmaker) [reply gremlin-go/driver/client.go]: The Python implementation had a comment that stated
```
# prevent the Client from being closed more than once. it raises errors if new jobby jobs
# get submitted to the executor when it is shutdown
```
would tracking closed be overkill for just this?

File: gremlin-go/driver/driverRemoteConnection.go (hunk 21 above)
Comment (lyndonbauto) [reply gremlin-go/driver/driverRemoteConnection.go:148]: @simonz-bq What exactly do you mean by this?
> It doesn't seem like the session is actually being used by any clients, whether it be being created with one, or having CreateSession be invoked.

Client uses it as follows:
```
	if client.Session != "" {
		request.args["session"] = client.Session
		request.processor = "session"
	}
```

> The distinction between adding a session to an existing DRC and creating a new one with a session are different use cases but it's not very obvious what is for what, as you can do both. Creating a session from an existing DRC is also confusing because it creates an instance of the same type.

Agreed this is strange however this is what they do in most of the other drivers.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 22]
```diff
@@ -100,6 +123,41 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
...
+		sessionHandler = func(settings *DriverRemoteConnectionSettings) {
+			settings.Session = uuid.New().String()
+		}
+	}
+	connection, err := NewDriverRemoteConnection(driver.client.url, sessionHandler)
+	if err != nil {
+		return nil, err
+	}
+	driver.spawnedSessions = append(driver.spawnedSessions, connection)
+	return connection, nil
```
Comment (lyndonbauto) [inline gremlin-go/driver/driverRemoteConnection.go]:
```suggestion
	if len(sessionId) > 1 {
		return nil, errors.New("more than one Session ID specified. Cannot create Session with multiple UUIDs")
	} else if driver.isSessionBound() {
		return nil, errors.New("connection is already bound to a Session - child sessions are not allowed")
	}
	
	driver.client.logHandler.logger.Log(Info, "creating Session based connection")
	sid := uuid.New().String()
	// If there is a sessionId provided, overwrite uuid.
         if len(sessionId) == 1 {
        	sid =  sessionId[0]
	}
	drc, err := NewDriverRemoteConnection(driver.client.url, func(settings *DriverRemoteConnectionSettings) {
			settings.Session = sid
		})
	if err != nil {
		return nil, err
	}
	driver.spawnedSessions = append(driver.spawnedSessions, drc)
	return drc, nil
```

File: gremlin-go/driver/client.go [hunk 23]
```diff
@@ -42,6 +43,7 @@ type Client struct {
 	logHandler      *logHandler
 	transporterType TransporterType
 	connection      *connection
+	Session         string
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: this shouldn't be exported I think.

File: gremlin-go/driver/client.go [hunk 24]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
 func (client *Client) Close() error {
-	return client.connection.close()
+	// If it is a Session, call closeSession
+	if client.Session != "" {
+		_, err := client.closeSession()
+		if err != nil {
+			return err
+		}
+	}
+	client.logHandler.logf(Info, "Closing Client with url %s", client.url)
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: I still don't think this works. Refer to existing usages of the logger for examples.

File: gremlin-go/driver/client.go [hunk 25]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
 // Close closes the client via connection.
 func (client *Client) Close() error {
-	return client.connection.close()
+	// If it is a Session, call closeSession
+	if client.Session != "" {
+		_, err := client.closeSession()
+		if err != nil {
+			return err
+		}
+	}
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go:89]: Should this be an if - else?

File: gremlin-go/driver/client.go [hunk 26]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+	client.logHandler.logf(Info, "Closing Client with url %s", client.url)
+	err := client.connection.close()
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: In this last step we can omit the `if err != nil` check and just return `client.connection.close()`

File: gremlin-go/driver/client.go [hunk 27]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
 	client.logHandler.logf(Debug, submitStartedBytecode, *bytecode)
 	request := makeBytecodeRequest(bytecode, client.traversalSource)
+	if client.Session != "" {
+		request.args["session"] = client.Session
+		request.processor = "session"
+	}
 	return client.connection.write(&request)
 }
+
+func (client *Client) closeSession() (ResultSet, error) {
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: I don't think this should return a `ResultSet`

File: gremlin-go/driver/client.go [hunk 28]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
 // Submit submits a Gremlin script to the server and returns a ResultSet.
 func (client *Client) Submit(traversalString string) (ResultSet, error) {
-	// TODO AN-982: Obtain connection from pool of connections held by the client.
 	client.logHandler.logf(Debug, submitStartedString, traversalString)
 	request := makeStringRequest(traversalString, client.traversalSource)
+	// TODO: Add bindings to request. request.args['bindings'] = bindings
+	if client.Session != "" {
+		request.args["session"] = client.Session
+		request.processor = "session"
+	}
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: Since this is shared functionality for modifying a request as below, let's make a helper method in `request.go`

File: gremlin-go/driver/driverRemoteConnection.go [hunk 29]
```diff
@@ -80,13 +85,36 @@ func NewDriverRemoteConnection(
...
+			}
+		}
+		if flag {
+			return errors.New("there was at least one error while closing spawned sessions. Check error log for more info")
+		}
+		return nil
+	}
+
+	if driver.isSessionBound() {
+		driver.client.logHandler.logf(Info, closeDRCSession, driver.client.url, driver.client.Session)
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: Abbreviations are typically against go best practices for naming.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 30]
```diff
@@ -80,13 +85,36 @@ func NewDriverRemoteConnection(
...
+	} else {
+		driver.client.logHandler.logf(Info, closeDRC, driver.client.url)
+	}
 	return driver.client.Close()
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: 1. If the driver has spawned sessions, then I'm assuming that `isSessionBound` check at the bottom here is supposed to trigger, which it will not because the function is returning in the loop above.
2. Since apparently your DRC can be live as well as have spawned sessions, then perhaps missing `driver.client.Close()` in the case where spawned sessions exist is a bug.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 31]
```diff
@@ -100,6 +128,39 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
 	return driver.client.submitBytecode(bytecode)
 }
 
+// isSessionBound returns True if a DriverRemoteConnection is a Session
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: Nit: can remove this comment off an internal method - also missing a period

File: gremlin-go/driver/driverRemoteConnection.go [hunk 32]
```diff
@@ -100,6 +128,39 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
...
+
+// CreateSession generates a new Session. sessionId stores the optional UUID param. It can be used to create a Session with a specific UUID.
+func (driver *DriverRemoteConnection) CreateSession(sessionId ...string) (*DriverRemoteConnection, error) {
+	if len(sessionId) > 1 {
+		return nil, errors.New("more than one Session ID specified. Cannot create Session with multiple UUIDs")
+	} else if driver.isSessionBound() {
+		return nil, errors.New("connection is already bound to a Session - child sessions are not allowed")
+	}
+
+	driver.client.logHandler.logger.Log(Info, "creating Session based connection")
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: Logging issue here again.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 33]
```diff
@@ -100,6 +128,39 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
...
+	sid := uuid.New().String()
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: This naming isn't the greatest and isn't necessary. You can just assign it directly in your func using an if statement.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/client.go (hunk 25 above)
Comment (L0Lmaker) [reply gremlin-go/driver/client.go:89]: No, we just need to perform a few extra steps if it's a session

File: gremlin-go/driver/driverRemoteConnection.go (hunk 30 above)
Comment (L0Lmaker) [reply gremlin-go/driver/driverRemoteConnection.go]: `isSessionBound` returns `True` if the DRC is a session, not if it has spawnedSessions. Perhaps I should rename it to just `isSession`

File: gremlin-go/driver/client.go (hunk 25 above)
Comment (simonz-bq) [reply gremlin-go/driver/client.go:89]: Hmm then I guess the issue here is that if there is an error, the second part will not run despite it being a separate close.

File: gremlin-go/driver/client.go [hunk 34]
```diff
@@ -84,12 +95,21 @@ func (client *Client) Submit(traversalString string) (ResultSet, error) {
 	// TODO AN-982: Obtain connection from pool of connections held by the client.
 	client.logHandler.logf(Debug, submitStartedString, traversalString)
 	request := makeStringRequest(traversalString, client.traversalSource)
+	// TODO: Add bindings to request. request.args['bindings'] = bindings
+	addSessionArgsToRequest(client, request)
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: 1. This should not take in the client. The `if` check needs to still be here.
2. You're passing the request by value in `addSessionArgsToRequest`, so modifying it won't work

File: gremlin-go/driver/driverRemoteConnection.go (hunk 30 above)
Comment (simonz-bq) [reply gremlin-go/driver/driverRemoteConnection.go]: Regardless, the client is not closing if there are spawned sessions.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/client.go [hunk 35]
```diff
@@ -71,11 +73,20 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+	// If it is a Session, call closeSession
+	if client.session != "" {
+		err := client.closeSession()
+		if err != nil {
+			return err
+		}
+	}
+	client.logHandler.logf(Info, closeClient, client.url)
 	return client.connection.close()
 }
```
Comment (lyndonbauto) [inline gremlin-go/driver/client.go]: You probably need something like this @L0Lmaker 

```suggestion
func (client *Client) Close() error {
	// If it is a Session, call closeSession
	var sessionError error
	if client.session != "" {
		sessionError = client.closeSession()
	}
	client.logHandler.logf(Info, closeClient, client.url)
	closeError := client.connection.close()
	// TODO: Combine sessionError and closeError  and return if both are not nil.
}
```

File: gremlin-go/driver/logger.go [hunk 36]
```diff
@@ -55,12 +55,12 @@ type defaultLogger struct {
...
-func (logger *defaultLogger) Log(verbosity LogVerbosity, v ...interface{}) {
+func (logger *defaultLogger) Log(_ LogVerbosity, v ...interface{}) {
 	log.Print(v...)
 }
 
 // Logf writes a formatted message to the defaultLogger.
-func (logger *defaultLogger) Logf(verbosity LogVerbosity, format string, v ...interface{}) {
+func (logger *defaultLogger) Logf(_ LogVerbosity, format string, v ...interface{}) {
 	log.Printf(format, v...)
 }
```
Comment (lyndonbauto) [inline gremlin-go/driver/logger.go:65]: @simonz-bq I just noticed we are ignoring verbosity, is this intentional?

File: gremlin-go/driver/client.go [hunk 37]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
 		transporterType: settings.TransporterType,
 		connection:      conn,
 	}
+	// TODO: PoolSize must be 1 on Session mode
 	return client, nil
 }
 
 // Close closes the client via connection.
 func (client *Client) Close() error {
```
Comment (lyndonbauto) [inline gremlin-go/driver/client.go]: Remove error from return signature and just return at end.

Comment (L0Lmaker) [reply gremlin-go/driver/client.go]: yep just pushing that now

**Review by lyndonbauto — APPROVED** [review]

File: gremlin-go/driver/client.go [hunk 38]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
 
 // Close closes the client via connection.
-func (client *Client) Close() error {
-	return client.connection.close()
+func (client *Client) Close() {
+	// If it is a Session, call closeSession
+	if client.session != "" {
+		err := client.closeSession()
+		if err != nil {
+			client.logHandler.logf(Error, closeSessionRequestError, client.url, client.session, err.Error())
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: Warning or debug is more appropriate here.

File: gremlin-go/driver/client.go [hunk 39]
```diff
@@ -71,25 +73,51 @@ func NewClient(url string, configurations ...func(settings *ClientSettings)) (*C
...
+		}
+		client.session = ""
+	}
+	client.logHandler.logf(Info, closeClient, client.url)
+	err := client.connection.close()
+	if err != nil {
+		client.logHandler.logf(Error, closeClientError, err.Error())
```
Comment (simonz-bq) [inline gremlin-go/driver/client.go]: Warning or debug is more appropriate here.

File: gremlin-go/driver/request.go [hunk 40]
```diff
@@ -75,3 +75,19 @@ func makeBasicAuthRequest(auth string) (req request) {
...
+		requestID: uuid.New(),
+		op:        "close",
+		processor: "session",
+		args: map[string]interface{}{
+			"session": client.session,
+		},
+	}
+}
+
+func addSessionArgsToRequest(request *request, sessionId string) {
```
Comment (simonz-bq) [inline gremlin-go/driver/request.go]: Since this mutates the request which is not aligned with how the rest of the requests are created, why not add the ability to add the sessionId to `makeByteCode` and `makeStringRequest`

File: gremlin-go/driver/logger.go (hunk 36 above)
Comment (simonz-bq) [reply gremlin-go/driver/logger.go:65]: Yes.

The `logHandler` deals with deciding whether or not to pass in the severity of the message.

The interface is just so anyone that implements their own logger has access to the verbosity if they want to do something with it.

File: gremlin-go/driver/driverRemoteConnection.go [hunk 41]
```diff
@@ -100,6 +120,37 @@ func (driver *DriverRemoteConnection) submitBytecode(bytecode *bytecode) (Result
...
+		return nil, errors.New("more than one Session ID specified. Cannot create Session with multiple UUIDs")
+	} else if driver.isSession() {
+		return nil, errors.New("connection is already bound to a Session - child sessions are not allowed")
+	}
+
+	driver.client.logHandler.log(Info, creatingSessionConnection)
+	drc, err := NewDriverRemoteConnection(driver.client.url, func(settings *DriverRemoteConnectionSettings) {
+		if len(sessionId) == 1 {
+			settings.Session = sessionId[0]
+			return
```
Comment (simonz-bq) [inline gremlin-go/driver/driverRemoteConnection.go]: An `if else` makes more sense than a return here.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/request.go [hunk 42]
```diff
@@ -75,3 +75,19 @@ func makeBasicAuthRequest(auth string) (req request) {
 		},
 	}
 }
+
+func makeCloseSessionRequest(client *Client) request {
```
Comment (simonz-bq) [inline gremlin-go/driver/request.go]: This doesn't need the entire client - just the session is fine.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

**Review by simonz-bq — APPROVED** [review]
